### Eclipse Workspace Patch 1.0
#P dwr
Index: java/org/directwebremoting/dwrp/BaseCallMarshaller.java
===================================================================
RCS file: /cvs/dwr/java/org/directwebremoting/dwrp/BaseCallMarshaller.java,v
retrieving revision 1.47
diff -u -r1.47 BaseCallMarshaller.java
--- java/org/directwebremoting/dwrp/BaseCallMarshaller.java	20 Feb 2008 14:52:54 -0000	1.47
+++ java/org/directwebremoting/dwrp/BaseCallMarshaller.java	27 Feb 2008 02:21:41 -0000
@@ -463,6 +463,8 @@
         String script = EnginePrivate.getRemoteHandleBatchExceptionScript(batchId, ex);
         out.print(script);
         sendOutboundScriptSuffix(out, batchId);
+
+        log.debug("Error processing request", ex);
     }
 
     /**
